Add scenario tests to the application

sebastian-quintero opened the pull request:
# Description

Adds new functionality for the `cloud.Application`:

- `new_scenario_test`: create a new scenario test.
- `scenario_test`: get a scenario test.
- `list_scenario_tests`: list scenario tests.
- `delete_scenario_tests`: deletes a scenario test.

sebastian-quintero committed:
Towards scenario tests

sebastian-quintero committed:
Adds scenario testing to the cloud client

This commit introduces scenario testing capabilities to the cloud client, allowing users to test their applications with various scenarios.

It adds the ability to create input sets from various sources, including runs, managed inputs, and new data. It also adds the ability to create and manage managed inputs, which are inputs that are stored in the cloud.

sebastian-quintero committed:
Adds support for scenario tests

Adds functionality to manage scenario tests, which are based on batch experiments.

Implements methods to create, list, delete, and retrieve scenario tests,
allowing users to manage and analyze test scenarios in the cloud platform.

merschformann commented on nextmv/nextmv/cloud/application.py:
```diff
@@ -230,18 +241,40 @@ def delete_batch_experiment(self, batch_id: str) -> None:
+            If the response status code is not 2xx.
+        """
+
+        _ = self.delete_batch_experiment(batch_id=scenario_test_id)
```
Why is it btw that we are explicitly throwing away the return value of some functions (`_ = ...`). This one does not even have a return value, right? (it's a `-> None` I think)
Is it a style convention? :thinking:

sebastian-quintero replied:
No convention here, just me being stupid. I fixed it.

merschformann commented on nextmv/nextmv/cloud/application.py:
```diff
@@ -409,6 +447,53 @@ def list_instances(self) -> list[Instance]:
+
+        response = self.client.request(
+            method="GET",
+            endpoint=f"{self.experiments_endpoint}/batch?",
```
Is the trailing `?` necessary here? The underlying `requests` should automatically add it when adding params, right? :thinking: 
Not important though if it works. It just looks a bit odd to me.

sebastian-quintero replied:
Leftover from some stuff I was trying. Nice catch, I removed it.

merschformann commented on nextmv/nextmv/cloud/application.py:
```diff
@@ -1048,6 +1297,124 @@ def new_run_with_result(
+        scenarios_by_id = {}
+        instances = {}
+        for scenario_ix, scenario in enumerate(scenarios):
+            scenario_id = f"scenario-{scenario_ix}" if scenario.scenario_id is None else scenario.scenario_id
```
Do we need to verify that we are generating a unique ID here? I.e., `scenario.scenario_id` of one scenario could technically already be defined as "scenario-0" which we would generate for the first if it wouldn't have an ID. :thinking: 

Probably a bit overcautious for now. I only use scenario testing occasionally and don't have a good sense for the use case of this `nextmv-py` functionality yet. :blush:

sebastian-quintero replied:
Good point. Better be safe than sorry. I will create a function for handling scenario IDs creation and checking that they result in unique keys.

sebastian-quintero replied:
I created the `_scenarios_by_id` that creates the IDs of the scenarios and makes sure that they are unique.

merschformann commented on nextmv/nextmv/cloud/run.py:
```diff
@@ -164,7 +164,7 @@ class RunConfiguration(BaseModel):
     """ID of the secrets collection to use for the run."""
-    queuing: Optional[RunQueueing] = None
+    queuing: Optional[RunQueuing] = None
     """Queueing configuration for the run."""
```
```suggestion
    """Queuing configuration for the run."""
```

Here is another one. :blush:

sebastian-quintero replied:
Thanks

merschformann commented on nextmv/nextmv/cloud/scenario.py:
```diff
@@ -0,0 +1,206 @@
+    """
+    Configuration for a scenario.
+
+    You can define multiple values for a single option, which will result in
+    multiple scenarios being created. For example, if you have a configuration
```
> You can define multiple values for a single option, which will result in **_multiple scenarios_** being created

I am actually not fully clear on what exactly we refer to as a "scenario". It seems like console UI may have a different understanding than what this comment states. This comment seems to suggest that any fine-granular / atomic / cannot be broken down further unit is a scenario.

I actually agree with that description, but am not sure whether we apply the terminology consistently. :thinking:

sebastian-quintero replied:
Good catch, I meant to say "runs" here, not scenario/

merschformann reviewed (approved):
Nice work! :hugs: 

I only left some minor comments and thoughts. Nothing really needs to be done for these. :blush:

sebastian-quintero committed:
Fix stuff that was wrong with scenario tests

sebastian-quintero committed:
Refactors scenario processing for efficiency

Improves scenario processing within applications by mapping scenarios to their IDs and preventing duplicate IDs. This refactoring enhances efficiency by creating scenario IDs when none is provided. It also ensures uniqueness, which is crucial for managing options effectively.
